### PR TITLE
``tile`` manipulation for cupy.ndarray

### PR DESCRIPTION
**Pull request by wkentaro**

### CODE REVIEW

**Comment by wkentaro** [comment]
For #599

File: cupy/manipulation/tiling.py [hunk 1]
```diff
@@ -1,7 +1,86 @@
...
+
+    .. seealso:: :func:`numpy.tile`
+
+    """
+    try:
+        tup = tuple(reps)
+    except TypeError:
+        tup = (reps,)
+    d = len(tup)
+    if all(x == 1 for x in tup) and isinstance(A, cupy.ndarray):
```
Comment (okuta) [inline cupy/manipulation/tiling.py]: Hou about `tup.count(1) == len(tup)`?

File: cupy/manipulation/tiling.py [hunk 2]
```diff
@@ -1,7 +1,86 @@
...
+        # Fixes the problem that the function does not make a copy if A is a
+        # array and the repetitions are 1 in all dimensions
+        return cupy.array(A, copy=True, ndmin=d)
+    else:
+        # Note that no copy of zero-sized arrays is made. However since they
+        # have no data there is no risk of an inadvertent overwrite.
+        c = cupy.array(A, copy=False, ndmin=d)
+    if (d < c.ndim):
+        tup = (1,)*(c.ndim-d) + tup
```
Comment (okuta) [inline cupy/manipulation/tiling.py]: Please add space. `tup = (1,) * (c.ndim - d) + tup`.

File: cupy/manipulation/tiling.py [hunk 3]
```diff
@@ -1,7 +1,86 @@
...
+    shape_out = tuple(s*t for s, t in zip(c.shape, tup))
```
Comment (okuta) [inline cupy/manipulation/tiling.py]: Please add space.

File: cupy/manipulation/tiling.py [hunk 4]
```diff
@@ -1,7 +1,86 @@
...
+    else:
+        # Note that no copy of zero-sized arrays is made. However since they
+        # have no data there is no risk of an inadvertent overwrite.
+        c = cupy.array(A, copy=False, ndmin=d)
+    if (d < c.ndim):
+        tup = (1,) * (c.ndim-d) + tup
+    shape_out = tuple(s * t for s, t in zip(c.shape, tup))
+    n = c.size
+    if n > 0:
+        for dim_in, nrep in zip(c.shape, tup):
```
Comment (okuta) [inline cupy/manipulation/tiling.py]: How about use broadcast?
Maybe you can write a single assignment.

File: cupy/manipulation/tiling.py [hunk 5]
```diff
@@ -1,7 +1,86 @@
...
+            if nrep != 1:
+                c = c.reshape(-1, n).repeat(nrep, 0)
```
Comment (okuta) [inline cupy/manipulation/tiling.py]: How about use broadcast?
Maybe you can write a single assignment.

Comment (okuta) [inline cupy/manipulation/tiling.py]: For example,
``` python
if n == 0:
    return cupy.empty(shape_out, dtype=c.dtype)
c_shape = []
ret_shape = []
for dim_in, nrep in zip(c.shape, tup):
    if nrep == 1:
        c_shape.append(dim_in)
        ret_shape.append(dim_in)
    elif dim_in == 1:
        c_shape.append(dim_in)
        ret_shape.append(nrep)
    else:
        c_shape.append(1)
        c_shape.append(dim_in)
        ret_shape.append(nrep)
        ret_shape.append(dim_in)
ret = cupy.empty(ret_shape, dtype=c.dtype)
ret[...] = c.reshape(c_shape)
return ret.reshape(shape_out)
```

**Comment by wkentaro** [comment]
I have done.

File: cupy/manipulation/tiling.py [hunk 6]
```diff
@@ -1,7 +1,117 @@
...
+        return cupy.array(A, copy=True, ndmin=d)
+    else:
+        # Note that no copy of zero-sized arrays is made. However since they
+        # have no data there is no risk of an inadvertent overwrite.
+        c = cupy.array(A, copy=False, ndmin=d)
+    if (d < c.ndim):
+        tup = (1,) * (c.ndim-d) + tup
+    shape_out = tuple(s * t for s, t in zip(c.shape, tup))
+    n = c.size
+    if n == 0:
```
Comment (okuta) [inline cupy/manipulation/tiling.py]: How about `c.size`?

File: cupy/manipulation/tiling.py [hunk 7]
```diff
@@ -1,7 +1,117 @@
...
+    if tup.count(1) == len(tup) and isinstance(A, cupy.ndarray):
+        # Fixes the problem that the function does not make a copy if A is a
+        # array and the repetitions are 1 in all dimensions
+        return cupy.array(A, copy=True, ndmin=d)
+    else:
+        # Note that no copy of zero-sized arrays is made. However since they
+        # have no data there is no risk of an inadvertent overwrite.
+        c = cupy.array(A, copy=False, ndmin=d)
+    if (d < c.ndim):
+        tup = (1,) * (c.ndim-d) + tup
```
Comment (okuta) [inline cupy/manipulation/tiling.py]: Please add space. `(c.ndim - d)`

**Comment by wkentaro** [comment]
Updated.

**Comment by okuta** [comment]
Please add test of changing the conditions of resp.
- zero: e.g. `resp=0`
- one: e.g. `resp=1`
- negative: e.g. `resp=-1`
- long: e.g. `resp=(2, 3, 4, 5)`

**Comment by wkentaro** [comment]
enough? https://github.com/pfnet/chainer/pull/605/files#diff-075b672822ebbfe27365fdc6e53d5b36R44

**Comment by okuta** [comment]
Perfect!
LGTM.

**Comment by wkentaro** [comment]
Thanks!

**Pull request merged**